Add initial docs for patroni --validate/generate config (#2929)
zalando/patroni: docs/patroni_configuration.rst

Upon changing these options, Patroni will read the relevant section of the configuration stored in DCS and change its run-time values.

Patroni nodes are dumping the state of the DCS options to disk upon for every change of the configuration into the file ``patroni.dynamic.json`` located in the Postgres data directory. Only the leader is allowed to restore these options from the on-disk dump if these are completely absent from the DCS or if they are invalid.

.. _validate_generate_config:

Configuration generation and validation
---------------------------------------

Patroni provides command-line interfaces for a Patroni :ref:`local configuration <yaml_configuration>` generation and validation. Using the ``patroni`` executable you can:

- Create a sample local Patroni configuration;

> - Create a local Patroni configuration for the locally running PostgreSQL instance (e.g. as a preparation step for the :ref:`Patroni integration <existing_data>`);

Reviewer: ```suggestion
- Create a Patroni configuration file for the locally running PostgreSQL instance (e.g. as a preparation step for the :ref:`Patroni integration <existing_data>`);
```

> - Validate a given local Patroni configuration.

Reviewer: ```suggestion
- Validate a given Patroni configuration file.
```

Sample Patroni configuration
^^^^^^^^^^^^^^^^^^^^^^^^^^^^

.. code:: text

   patroni --generate-sample-config [configfile]

Description
"""""""""""

> Generate a sample Patroni local configuration in ``yaml`` format.

Reviewer: ```suggestion
Generate a sample Patroni configuration file in ``yaml`` format.
```

> Parameter values are defined using the :ref:`Environment configuration <environment>`, otherwise, if not set, the defaults used in Patroni or to the '#FIXME' string for the values that should be later defined by the user.

Reviewer 2: ```suggestion
Parameter values are defined using the :ref:`Environment configuration <environment>`, otherwise, if not set, the defaults used in Patroni or the ``#FIXME`` string for the values that should be later defined by the user.
```

Some default values are defined based on the local setup:

>    -  **postgresql.listen**: the IP address returned by ``gethostname`` call for the current machine's hostname and the standart ``5432`` port.
>    -  **postgresql.connect_address**: the IP address returned by ``gethostname`` call for the current machine's hostname and the standart ``5432`` port.
>    -  **postgresql.authentication.rewind**: is only defined if the PostgreSQL version can be defined from the binary and the version is 11 or later.
>    -  **restapi.listen**: IP address returned by ``gethostname`` call for the current machine's hostname and the standart ``8008`` port.
>    -  **restapi.connect_address**: IP address returned by ``gethostname`` call for the current machine's hostname and the standart ``8008`` port.

Reviewer 2: ```suggestion
   -  **postgresql.listen**: the IP address returned by ``gethostname`` call for the current machine's hostname and the standard ``5432`` port.
   -  **postgresql.connect_address**: the IP address returned by ``gethostname`` call for the current machine's hostname and the standard ``5432`` port.
   -  **postgresql.authentication.rewind**: is only defined if the PostgreSQL version can be defined from the binary and the version is 11 or later.
   -  **restapi.listen**: IP address returned by ``gethostname`` call for the current machine's hostname and the standard ``8008`` port.
   -  **restapi.connect_address**: IP address returned by ``gethostname`` call for the current machine's hostname and the standard ``8008`` port.
```

Parameters
""""""""""

``configfile`` - full path to the configuration file used to store the result. If not provided, the result is sent to ``stdout``.

Patroni configuration for a running instance
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

.. code:: text

   patroni --generate-config [--dsn DSN] [configfile]

Description
"""""""""""

> Generate Patroni local configuration in ``yaml`` format for the locally running PostgreSQL instance. 

Reviewer 2: ```suggestion
Generate a Patroni configuration in ``yaml`` format for the locally running PostgreSQL instance. 
```

> Either the provided DSN (takes precedence) or PostgreSQL environment variables will be used for the PostgreSQL connection. If password is not provided, it should be entered via prompt.

Reviewer: ```suggestion
Either the provided DSN (takes precedence) or PostgreSQL `environment variables <https://www.postgresql.org/docs/current/libpq-envars.html>`__ will be used for the PostgreSQL connection. If password is not provided, it should be entered via prompt.
```

> Either the provided DSN (takes precedence) or PostgreSQL `environment variables <https://www.postgresql.org/docs/current/libpq-envars.html>`__ will be used for the PostgreSQL connection. If password is not provided, it should be entered via prompt.

Reviewer 2: ```suggestion
Either the provided DSN (takes precedence) or PostgreSQL `environment variables <https://www.postgresql.org/docs/current/libpq-envars.html>`__ will be used for the PostgreSQL connection. If the password is not provided, it should be entered via prompt.
```

> All the non-internal GUCs having configuration file, postmaster command line or environment variable as a source defined on the running instance will be used for the following Patroni configuration parameters:

Reviewer 2: ```suggestion
All the non-internal GUCs defined in the source Postgres instance, independently if they were set through a configuration file, through the postmaster command-line, or through environment variables, will be used as the source for the following Patroni configuration parameters:
```

   -  **scope**: ``cluster_name`` GUC value;
   -  **postgresql.listen**: ``listen_addresses`` and ``port`` GUC values;
   -  **postgresql.datadir**: ``data_directory`` GUC value;
   -  **postgresql.parameters**: ``archive_command``, ``restore_command``, ``archive_cleanup_command``, ``recovery_end_command``, ``ssl_passphrase_command``, ``hba_file``, ``ident_file``, ``config_file`` GUC values;
   -  **bootstrap.dcs**: all other gathered PostgreSQL GUCs.

> If ``scope``, ``postgresql.listen`` or ``postgresql.datadir`` is not set form the GUCs, the respective :ref:`Environment configuration <environment>` value is used.

Reviewer 2: ```suggestion
If ``scope``, ``postgresql.listen`` or ``postgresql.datadir`` is not set from the Postgres GUCs, the respective :ref:`Environment configuration <environment>` value is used.
```

Other rules applied for the values definition:

   -  **name**: ``PATRONI_NAME`` environment variable value if set, otherwise the current machine's hostname.
   -  **postgresql.bin_dir**: path to the Postgres binaries gathered from the running instance.
   -  **postgresql.connect_address**: the IP address returned by ``gethostname`` call for the current machine's hostname and the port used for the instance connection or the ``port`` GUC value.
   -  **postgresql.authentication.superuser**: the configuration used for the instance connection;
   -  **postgresql.pg_hba**: the lines gathered from the source instance's ``hba_file``.
   -  **postgresql.pg_ident**: the lines gathered from the source instance's ``ident_file``.

>    -  **restapi.listen**: IP address returned by ``gethostname`` call for the current machine's hostname and the standart ``8008`` port.
>    -  **restapi.connect_address**: IP address returned by ``gethostname`` call for the current machine's hostname and the standart ``8008`` port.

Reviewer 2: ```suggestion
   -  **restapi.listen**: IP address returned by ``gethostname`` call for the current machine's hostname and the standard ``8008`` port.
   -  **restapi.connect_address**: IP address returned by ``gethostname`` call for the current machine's hostname and the standard ``8008`` port.
```

> Other parameters defined using :ref:`Environment configuration <environment>` are also included into the configurtion.

Reviewer 2: ```suggestion
Other parameters defined using :ref:`Environment configuration <environment>` are also included into the configuration.
```

Parameters
""""""""""

``configfile``
    Full path to the configuration file used to store the result. If not provided, result is sent to ``stdout``.

``dsn``
    Optional DSN string for the local PostgreSQL instance to get GUC values from.

Validate Patroni configuration
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

.. code:: text

   patroni --validate-config [configfile]

Description
"""""""""""

> Validate the given Patroni local configuration and print the information about the failed checks.

Reviewer 2: ```suggestion
Validate the given Patroni configuration and print the information about the failed checks.
```
